tests/types/test_dsl.py

```diff
@@ -254,3 +278,270 @@ def test_dsl_display():
         tree
         == "└── Sequence\n    ├── KleeneStar(*)\n    │   └── Alternatives(|)\n    │       ├── String('a')\n    │       └── String('b')\n    └── Regex('[0-9]')\n"
     )
+
+
+def test_dsl_python_types_to_terms():
+    with pytest.raises(RecursionError):
+        python_types_to_terms(None, 11)
+
+    term = Term()
+    assert python_types_to_terms(term) == term
+
+    assert python_types_to_terms(int) == types.integer
+    assert python_types_to_terms(float) == types.number
+    assert python_types_to_terms(bool) == types.boolean
+    assert python_types_to_terms(str) == types.string
+    assert python_types_to_terms(datetime.time) == types.time
+    assert python_types_to_terms(datetime.date) == types.date
+    assert python_types_to_terms(datetime.datetime) == types.datetime
+    assert python_types_to_terms(dict) == types.CFG(grammars.json)
+
+    string_instance = "a"
+    assert python_types_to_terms(string_instance) == String(string_instance)
+    int_instance = 1
+    assert python_types_to_terms(int_instance) == Regex(r"1")
+    float_instance = 1.0
+    assert python_types_to_terms(float_instance) == Regex(r"1.0")
+
+    @dataclass
+    class DataClass:
+        a: int
+        b: str
+
+    assert python_types_to_terms(DataClass) == JsonSchema(
+        {
+            "properties": {"a": {"title": "A", "type": "integer"}, "b": {"title": "B", "type": "string"}},
+            "required": ["a", "b"],
+            "title": "DataClass",
+            "type": "object",
+        }
+    )
+
+    class SomeTypedDict(TypedDict):
+        a: int
+        b: str
+
+    assert python_types_to_terms(SomeTypedDict) == JsonSchema(
+        {
+            "properties": {"a": {"title": "A", "type": "integer"}, "b": {"title": "B", "type": "string"}},
+            "required": ["a", "b"],
+            "title": "SomeTypedDict",
+            "type": "object",
+        }
+    )
+
+    class PydanticModel(BaseModel):
+        a: int
+        b: str
+
+    assert python_types_to_terms(PydanticModel) == JsonSchema(
+        {
+            "properties": {"a": {"title": "A", "type": "integer"}, "b": {"title": "B", "type": "string"}},
+            "required": ["a", "b"],
+            "title": "PydanticModel",
+            "type": "object",
+        }
+    )
+
+    builder = SchemaBuilder()
+    builder.add_schema({"type": "object", "properties": {}})
+    builder.add_object({"hi": "there"})
+    builder.add_object({"hi": 5})
+    assert python_types_to_terms(builder) == JsonSchema(
+        {
+            "$schema": "http://json-schema.org/schema#",
+            "type": "object",
+            "properties": {"hi": {"type": ["integer", "string"]}},
+            "required": ["hi"]
+        }
+    )
+
+    interegular_fsm = interegular.fsm.FSM(
+        alphabet={},
+        states={},
+        initial={},
+        finals={},
+        map={},
+        __no_validation__=True,
+    )
+    assert python_types_to_terms(interegular_fsm) is interegular_fsm
```

Comment:
I don't think this works with `RegexGuide` anymore, in which case we'd have to remove it

Reply:
It still works in the sense that it will generate text that follows the FSM as expected. Maybe we can discuss what we do with it as part of this issue: https://github.com/dottxt-ai/outlines/issues/1487